From 7873f8334c8d31031f8cfa83bd97ac6029309e4f Mon Sep 17 00:00:00 2001
From: "Todd C. Miller" <Todd.Miller@sudo.ws>
Date: Sat, 9 Sep 2023 14:07:04 -0600
Subject: [PATCH] Try to make sudo less vulnerable to ROWHAMMER attacks.

We now use ROWHAMMER-resistent values for ALLOW, DENY, AUTH_SUCCESS,
AUTH_FAILURE, AUTH_ERROR and AUTH_NONINTERACTIVE.  In addition, we
explicitly test for expected values instead of using a negated test
against an error value.  In the parser match functions this means
explicitly checking for ALLOW or DENY instead of accepting anything
that is not set to UNSPEC.

Thanks to Andrew J. Adiletta, M. Caner Tol, Yarkin Doroz, and Berk
Sunar, all affiliated with the Vernam Applied Cryptography and
Cybersecurity Lab at Worcester Polytechnic Institute, for the report.
Paper preprint: https://arxiv.org/abs/2309.02545
---
 plugins/sudoers/auth/passwd.c    | 27 ++++++++++-------
 

Index: sudo-1.9.13p3/plugins/sudoers/auth/passwd.c
===================================================================
--- sudo-1.9.13p3.orig/plugins/sudoers/auth/passwd.c
+++ sudo-1.9.13p3/plugins/sudoers/auth/passwd.c
@@ -66,7 +66,7 @@ sudo_passwd_verify(struct passwd *pw, co
     char des_pass[9], *epass;
     char *pw_epasswd = auth->data;
     size_t pw_len;
-    int matched = 0;
+    int ret;
     debug_decl(sudo_passwd_verify, SUDOERS_DEBUG_AUTH);
 
     /* An empty plain-text password must match an empty encrypted password. */
@@ -78,7 +78,7 @@ sudo_passwd_verify(struct passwd *pw, co
      */
     pw_len = strlen(pw_epasswd);
     if (pw_len == DESLEN || HAS_AGEINFO(pw_epasswd, pw_len)) {
-	strlcpy(des_pass, pass, sizeof(des_pass));
+	(void)strlcpy(des_pass, pass, sizeof(des_pass));
 	pass = des_pass;
     }
 
@@ -88,29 +88,36 @@ sudo_passwd_verify(struct passwd *pw, co
      * only compare the first DESLEN characters in that case.
      */
     epass = (char *) crypt(pass, pw_epasswd);
+    ret = AUTH_FAILURE;
     if (epass != NULL) {
-	if (HAS_AGEINFO(pw_epasswd, pw_len) && strlen(epass) == DESLEN)
-	    matched = !strncmp(pw_epasswd, epass, DESLEN);
-	else
-	    matched = !strcmp(pw_epasswd, epass);
+	if (HAS_AGEINFO(pw_epasswd, pw_len) && strlen(epass) == DESLEN) {
+	    if (strncmp(pw_epasswd, epass, DESLEN) == 0)
+		ret = AUTH_SUCCESS;
+	} else {
+	    if (strcmp(pw_epasswd, epass) == 0)
+		ret = AUTH_SUCCESS;
+	}
     }
 
     explicit_bzero(des_pass, sizeof(des_pass));
 
-    debug_return_int(matched ? AUTH_SUCCESS : AUTH_FAILURE);
+    debug_return_int(ret);
 }
 #else
 int
 sudo_passwd_verify(struct passwd *pw, const char *pass, sudo_auth *auth, struct sudo_conv_callback *callback)
 {
     char *pw_passwd = auth->data;
-    int matched;
+    int ret;
     debug_decl(sudo_passwd_verify, SUDOERS_DEBUG_AUTH);
 
     /* Simple string compare for systems without crypt(). */
-    matched = !strcmp(pass, pw_passwd);
+    if (strcmp(pass, pw_passwd) == 0)
+           ret = AUTH_SUCCESS;
+    else
+           ret = AUTH_FAILURE;
 
-    debug_return_int(matched ? AUTH_SUCCESS : AUTH_FAILURE);
+    debug_return_int(ret);
 }
 #endif
 
diff --git a/plugins/sudoers/regress/testsudoers/test2.sh b/plugins/sudoers/regress/testsudoers/test2.sh
index 0b0b3f8..b91dee5 100755
--- a/plugins/sudoers/regress/testsudoers/test2.sh
+++ b/plugins/sudoers/regress/testsudoers/test2.sh
@@ -5,6 +5,9 @@
 
 : ${TESTSUDOERS=testsudoers}
 
+chmod go-w $TESTDIR/test2.inc
+
+
 MYUID=`\ls -ln $TESTDIR/test2.inc | awk '{print $3}'`
 MYGID=`\ls -ln $TESTDIR/test2.inc | awk '{print $4}'`
 exec 2>&1
diff --git a/plugins/sudoers/regress/testsudoers/test8.sh b/plugins/sudoers/regress/testsudoers/test8.sh
index c22b590..619a89a 100755
--- a/plugins/sudoers/regress/testsudoers/test8.sh
+++ b/plugins/sudoers/regress/testsudoers/test8.sh
@@ -6,6 +6,9 @@
 
 : ${TESTSUDOERS=testsudoers}
 
+chmod go-w $TESTDIR/test2.inc
+
+
 MYUID=`\ls -ln $TESTDIR/test2.inc | awk '{print $3}'`
 MYGID=`\ls -ln $TESTDIR/test2.inc | awk '{print $4}'`
 exec 2>&1
diff --git a/plugins/sudoers/regress/testsudoers/test9.sh b/plugins/sudoers/regress/testsudoers/test9.sh
index 850bbac..8856b53 100755
--- a/plugins/sudoers/regress/testsudoers/test9.sh
+++ b/plugins/sudoers/regress/testsudoers/test9.sh
@@ -5,6 +5,9 @@
 
 : ${TESTSUDOERS=testsudoers}
 
+chmod go-w $TESTDIR/test2.inc
+
+
 MYUID=`\ls -ln $TESTDIR/test2.inc | awk '{print $3}'`
 MYGID=`\ls -ln $TESTDIR/test2.inc | awk '{print $4}'`
 exec 2>&1
From bc15ac5487151969048f369111f1745d38f096f4 Mon Sep 17 00:00:00 2001
Date: Thu, 17 Jul 2025 13:01:20 +0530
Subject: [PATCH 1/2] manual hunk chnages

---
 plugins/sudoers/auth/sudo_auth.c | 57 ++++++++++++++++++++++----------
 plugins/sudoers/auth/sudo_auth.h | 14 ++++----

Index: sudo-1.9.13p3/plugins/sudoers/auth/sudo_auth.c
===================================================================
--- sudo-1.9.13p3.orig/plugins/sudoers/auth/sudo_auth.c
+++ sudo-1.9.13p3/plugins/sudoers/auth/sudo_auth.c
@@ -115,10 +115,16 @@ sudo_auth_init(struct passwd *pw, int mo
 	if (auth->init && !IS_DISABLED(auth)) {
 	    /* Disable if it failed to init unless there was a fatal error. */
 	    status = (auth->init)(pw, auth);
-	    if (status == AUTH_FAILURE)
-		SET(auth->flags, FLAG_DISABLED);
-	    else if (status == AUTH_FATAL)
-		break;		/* assume error msg already printed */
+	    switch (status) {
+	    	case AUTH_SUCCESS:
+				break;
+	    	case AUTH_FAILURE:
+ 				SET(auth->flags, FLAG_DISABLED);
+				break;
+			default:
+				/* Assume error msg already printed. */
+				debug_return_int(-1);
+	    }
 	}
     }
 
@@ -164,7 +170,7 @@ sudo_auth_init(struct passwd *pw, int mo
 	}
     }
 
-    debug_return_int(status == AUTH_FATAL ? -1 : 0);
+    debug_return_int(0);
 }
 
 /*
@@ -205,7 +211,7 @@ sudo_auth_cleanup(struct passwd *pw, boo
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->cleanup && !IS_DISABLED(auth)) {
 	    int status = (auth->cleanup)(pw, auth, force);
-	    if (status == AUTH_FATAL) {
+	    if (status != AUTH_SUCCESS) {
 		/* Assume error msg already printed. */
 		debug_return_int(-1);
 	    }
@@ -302,7 +308,7 @@ verify_user(struct passwd *pw, char *pro
 		    SET(auth->flags, FLAG_DISABLED);
 		else if (status == AUTH_NONINTERACTIVE)
 		    goto done;
-		else if (status == AUTH_FATAL || user_interrupted())
+		else if (status != AUTH_SUCCESS || user_interrupted())
 		    goto done;		/* assume error msg already printed */
 	    }
 	}
@@ -360,7 +366,6 @@ done:
 	case AUTH_NONINTERACTIVE:
 	    SET(validated, FLAG_NO_USER_INPUT);
 	    FALLTHROUGH;
-	case AUTH_FATAL:
 	default:
 	    log_auth_failure(validated, 0);
 	    ret = -1;
@@ -378,18 +383,26 @@ int
 sudo_auth_begin_session(struct passwd *pw, char **user_env[])
 {
     sudo_auth *auth;
+	int ret = true;
     debug_decl(sudo_auth_begin_session, SUDOERS_DEBUG_AUTH);
 
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->begin_session && !IS_DISABLED(auth)) {
 	    int status = (auth->begin_session)(pw, user_env, auth);
-	    if (status != AUTH_SUCCESS) {
-		/* Assume error msg already printed. */
-		debug_return_int(-1);
-	    }
-	}
+	    switch (status) {
+	    	case AUTH_SUCCESS:
+				break;
+	    	case AUTH_FAILURE:
+				ret = false;
+				break;		
+		    default:
+				/* Assume error msg already printed. */
+				ret = -1;
+				break;
+	    	}
+		}	
     }
-    debug_return_int(1);
+    debug_return_int(ret);
 }
 
 bool
@@ -416,19 +429,27 @@ int
 sudo_auth_end_session(struct passwd *pw)
 {
     sudo_auth *auth;
+	int ret = true;
     int status;
     debug_decl(sudo_auth_end_session, SUDOERS_DEBUG_AUTH);
 
     for (auth = auth_switch; auth->name; auth++) {
 	if (auth->end_session && !IS_DISABLED(auth)) {
 	    status = (auth->end_session)(pw, auth);
-	    if (status == AUTH_FATAL) {
-		/* Assume error msg already printed. */
-		debug_return_int(-1);
+	    switch (status) {
+	    	case AUTH_SUCCESS:
+				break;
+	    	case AUTH_FAILURE:
+				ret = false;
+				break;
+	    	default:
+				/* Assume error msg already printed. */
+				ret = -1;
+				break;
 	    }
 	}
     }
-    debug_return_int(1);
+    debug_return_int(ret);
 }
 
 /*
Index: sudo-1.9.13p3/plugins/sudoers/auth/sudo_auth.h
===================================================================
--- sudo-1.9.13p3.orig/plugins/sudoers/auth/sudo_auth.h
+++ sudo-1.9.13p3/plugins/sudoers/auth/sudo_auth.h
@@ -19,12 +19,14 @@
 #ifndef SUDO_AUTH_H
 #define SUDO_AUTH_H
 
-/* Auth function return values.  */
-#define AUTH_SUCCESS		0
-#define AUTH_FAILURE		1
-#define AUTH_INTR		2
-#define AUTH_FATAL		3
-#define AUTH_NONINTERACTIVE	4
+#define AUTH_FATAL		0x1629e037	/* 0010110001010011110000000110111 */
+
+/* Auth function return values (rowhammer resistent).  */
+#define AUTH_SUCCESS		0x52a2925	/* 0101001010100010100100100101 */
+#define AUTH_FAILURE		0xad5d6da	/* 1010110101011101011011011010 */
+#define AUTH_INTR		0x69d61fc8	/* 1101001110101100001111111001000 */
+#define AUTH_ERROR		0x1629e037	/* 0010110001010011110000000110111 */
+#define AUTH_NONINTERACTIVE	0x1fc8d3ac	/* 11111110010001101001110101100 */
 
 typedef struct sudo_auth {
     int flags;			/* various flags, see below */
